Content.Server/Atmos/EntitySystems/GasAnalyzerSystem.cs

```diff
@@ -1,46 +1,276 @@
+            if (!_activeAnalyzers.Any())
+                return;
+
+            foreach (var analyzer in _activeAnalyzers)
             {
-                analyzer.Update(frameTime);
+                if (_invalidAnalyzers.List != null && _invalidAnalyzers.List.Contains(analyzer))
+                    continue;
+
+                if(!UpdateAnalyzer(analyzer))
+                    _invalidAnalyzers.Add(analyzer);
             }
+
+            foreach (var invalidAnalyzer in _invalidAnalyzers)
+            {
+                _activeAnalyzers.Remove(invalidAnalyzer);
+            }
+
+            _invalidAnalyzers.List?.Clear();
         }
 
+        /// <summary>
+        /// Activates the analyzer when used in the world, scanning either the target entity or the tile clicked
+        /// </summary>
         private void OnAfterInteract(EntityUid uid, GasAnalyzerComponent component, AfterInteractEvent args)
         {
             if (!args.CanReach)
             {
                 _popup.PopupEntity(Loc.GetString("gas-analyzer-component-player-cannot-reach-message"), args.User, Filter.Entities(args.User));
                 return;
             }
+            ActivateAnalyzer(uid, component, args.User, args.Target);
+            OpenUserInterface(args.User, component);
+            args.Handled = true;
+        }
+
+        /// <summary>
+        /// Activates the analyzer with no target, so it only scans the tile the user was on when activated
+        /// </summary>
+        private void OnUseInHand(EntityUid uid, GasAnalyzerComponent component, UseInHandEvent args)
+        {
+            ActivateAnalyzer(uid, component, args.User);
+            args.Handled = true;
+        }
+
+        /// <summary>
+        /// Handles analyzer activation logic
+        /// </summary>
+        private void ActivateAnalyzer(EntityUid uid, GasAnalyzerComponent component, EntityUid user, EntityUid? target = null)
+        {
+            component.Target = target;
+            component.User = user;
+            component.LastPosition = EntityManager.GetComponent<TransformComponent>(target ?? user).Coordinates;
+            component.Enabled = true;
+            Dirty(component);
+            SetAppearance(component);
+            _activeAnalyzers.Add(uid);
+        }
+
+        private void OnComponentShutdown(EntityUid uid, GasAnalyzerComponent component, ComponentShutdown args)
+        {
+            // Make sure to remove this analyzer from the update list
+            _activeAnalyzers.Remove(uid);
+        }
+
+        /// <summary>
+        /// Fetches new data when a user clicks on the refresh button
+        /// </summary>
+        private void OnRefreshMessage(EntityUid uid, GasAnalyzerComponent component, GasAnalyzerRefreshMessage message)
+        {
+            if (message.Session.AttachedEntity is not {Valid: true})
+                return;
+
+            UpdateAnalyzer(uid, component);
+        }
```

Comment:
having an explicit refresh button is pointless if the gas analyzer updates automatically isn't it? it also doesn't look great in the UI, so removing that would be nice

Reply:
Yeah this was a hold over from the prior implementation as it had a delay. But even when I add that delay, I agree, I'll remove it. Someone can always close/open the UI if for some reason it stops updating.